cache/cache.go

```diff
@@ -13,3 +13,288 @@
+		close(cache.ready)
+
+		serveWatchEvents(
+			internalCtx,
+			client, prefix,
+			cache.demux.broadcast,
+			cache.demux,
+			cache.history,
+			cfg.InitialBackoff,
+			cfg.MaxBackoff,
+			cfg.UpstreamBufferSize,
+		)
+	}()
+
+	return cache, nil
+}
+
+// Watch registers a cache-backed watcher for a given key or prefix.
+// It returns a WatchChan that streams WatchResponses containing events.
+func (c *Cache) Watch(ctx context.Context, key string, opts ...clientv3.OpOption) clientv3.WatchChan {
+	select {
+	case <-c.ready:
+	case <-ctx.Done():
+		emptyWatchChan := make(chan clientv3.WatchResponse)
+		close(emptyWatchChan)
+		return emptyWatchChan
+	}
+
+	if !strings.HasPrefix(key, c.prefix) {
+		emptyWatchChan := make(chan clientv3.WatchResponse)
+		close(emptyWatchChan)
+		return emptyWatchChan
+	}
+
+	op := clientv3.OpGet(key, opts...)
+	requestedRev := op.Rev()
+
+	responseChan := make(chan clientv3.WatchResponse, 1)
+
+	// build the internal Watch event stream (for historic reply + live)
+	stream, err := c.newWatchEventStream(ctx, key, requestedRev)
+	if err != nil {
+		switch err {
+		case ErrCompacted:
+			var compactRev int64
+			if oldestEvent := c.history.PeekOldest(); oldestEvent != nil {
+				compactRev = oldestEvent.Kv.ModRevision
+			}
+			responseChan <- clientv3.WatchResponse{CompactRevision: compactRev}
+		default:
+			responseChan <- clientv3.WatchResponse{Canceled: true}
+		}
+		close(responseChan)
+		return responseChan
+	}
+
+	go func() {
+		defer close(responseChan)
+		for {
+			select {
+			case event, ok := <-stream:
+				if !ok {
+					return
+				}
+				responseChan <- clientv3.WatchResponse{Events: []*clientv3.Event{event}}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return responseChan
+}
+
+// Ready reports whether the cache has finished its initial load.
+func (c *Cache) Ready() bool {
+	select {
+	case <-c.ready:
+		return true
+	default:
+		return false
+	}
+}
+
+// WaitReady blocks until the cache is ready or the ctx is cancelled.
+func (c *Cache) WaitReady(ctx context.Context) error {
+	select {
+	case <-c.ready:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Close cancels the private context and blocks until all goroutines return.
+func (c *Cache) Close() {
+	c.stop()
+	c.waitGroup.Wait()
+}
+
+// newWatchEventStream builds a single internal Watch event stream for
+// both historical replay (>= startRev) and live updates filtered by key.
+func (c *Cache) newWatchEventStream(ctx context.Context, key string, startRev int64) (WatchEventStream, error) {
+	pred := func(k []byte) bool { return strings.HasPrefix(string(k), key) }
+
+	// TODO: re-enable arbitrary startRev support once we guarantee gap-free replay.
+	if startRev != 0 {
+		return nil, ErrUnsupportedStartRev
+	}
+
+	lastSentRev := startRev - 1
+	if startRev == 0 {
+		if latestEvent := c.history.PeekLatest(); latestEvent != nil {
+			lastSentRev = latestEvent.Kv.ModRevision
+		}
+	}
+	nextRev := lastSentRev + 1
```

Comment:
In cases history is empty, the resulting nextRev will be `0`. Meaning you need to support `0` as special case. This means that this whole logic is not needed.